### PR TITLE
Add `runCommand` call

### PR DESCRIPTION
**Pull request by tylerkaraszewski**
### Details

Note: much easier to review with whitespace hidden.

In order to kill `acceptCommand` (which could have been called `enqueueCommand`) we need an alternative to call from auth's `forwardCommand` and Jobs's `_handleFailedRetryAfterQuery`.

This adds a `runCommand` that performs the core action from `::worker` (which worker also calls now) without queuing anything.

This PR is fairly significant in that it makes *most* (not all) commands run in socket threads rather than worker threads. This re-opens the possibility to remove the blocking command queue and allow every command to be blocking in it's own thread, without having the possibility of bottlenecking every other command by running out of workers.

### Fixed Issues
Fixes https://github.com/Expensify/Expensify/issues/221469

### Tests
Existing tests. Auth tested and passing against 8b951a66836a44a0445e8960228175aba521a525

### CODE REVIEW

File: BedrockServer.cpp [hunk 1]
```diff
@@ -713,385 +705,399 @@ void BedrockServer::worker(int threadId)
...
-                            // Move on to the next command until this one finishes.
-                            break;
-                        }
+                    if (!command->areHttpsRequestsComplete()) {
+                        // If it has outstanding HTTPS requests, we'll wait for them.
+                        waitForHTTPS(move(command));
+                    } else if (command->repeek) {
+                        // Otherwise, it needs to be re-peeked, but had no outstanding requests, so it goes
+                        // back in the main queue.
+                        _commandQueue.push(move(command));
```
Comment (rafecolton) [inline BedrockServer.cpp:974]: If I'm reading this correctly, this is a change in behavior - previously, `commandQueue` was set to either `_commandQueue` or `_blockingCommandQueue`, and now we're only using `_commandQueue`. Is this change intentional?

File: BedrockServer.cpp [hunk 2]
```diff
@@ -2256,45 +2259,47 @@ void BedrockServer::handleSocket(Socket&& socket, bool fromControlPort, bool fro
...
                         condition_variable cv;
+                        atomic<bool> finished = false;
 
-                        function<void()> callback = [&m, &cv]() {
+                        function<void()> callback = [&m, &cv, &finished]() {
                             // Lock the mutex above (which will be locked by this thread while we're queuing), which waits
                             // for `handleSocket` to release it's lock (by calling `wait`), and then notify the waiting
                             // socket thread.
-                            lock_guard lock(m);
+                            lock_guard l(m);
```
Comment (rafecolton) [inline BedrockServer.cpp:2269]: Is the variable rename necessary? Otherwise I would leave it to A) Reduce the diff; and B) Make it more readable. NAB though.

File: BedrockServer.cpp [hunk 3]
```diff
@@ -2256,45 +2259,47 @@ void BedrockServer::handleSocket(Socket&& socket, bool fromControlPort, bool fro
...
-                        // destructionCallback fires, this will stop blocking and we can move on to the next request.
-                        // NOTE: This doesn't correctly handle spurious wakeups.
-                        if (hasSocket) {
-                            cv.wait(lock);
+                        // Now that the command is queued, we wait for it to complete (if it's has a socket, and hasn't finished by the time we get to this point).
+                        // When this happens, destructionCallback fires, sets `finished` to true, and we can move on to the next request.
+                        unique_lock<mutex> lock(m);
+                        if (!finished && hasSocket) {
+                            cv.wait(lock, [&]{return finished.load();});
                         }
```
Comment (rafecolton) [inline BedrockServer.cpp:2303]: I see that `// NOTE: This doesn't correctly handle spurious wakeups.` was removed, so I assume that's what `finished` is fore. Can you walk me through the logic here?

File: BedrockServer.cpp [hunk 4]
```diff
@@ -2256,45 +2259,47 @@ void BedrockServer::handleSocket(Socket&& socket, bool fromControlPort, bool fro
...
                         auto _syncNodeCopy = atomic_load(&_syncNode);
                         if (_syncNodeCopy && _syncNodeCopy->getState() == SQLiteNode::STANDINGDOWN) {
                             _standDownQueue.push(move(command));
                         } else {
-                            SINFO("Queuing new '" << command->request.methodLine << "' command from local client, with "
-                                  << _commandQueue.size() << " commands already queued.");
-                            _commandQueue.push(move(command));
+                            // If this command is scheduled in the future, we can't just run it, we need to enqueue it to run at that point.
+                            // This functionality will go away as we remove the queues from bedrock, and so this can be removed at that time.
+                            if (command->request.calcU64("commandExecuteTime") > STimeNow()) {
```
Comment (rafecolton) [inline BedrockServer.cpp:2288]: Where was this handled before?

File: test/clustertest/tests/BadCommandTest.cpp [hunk 5]
```diff
@@ -14,6 +14,10 @@ struct BadCommandTest : tpunit::TestFixture {
                 BedrockTester& leader = tester.getTester(0);
                 BedrockTester& follower = tester.getTester(1);
 
+                ASSERT_TRUE(tester.getTester(0).waitForState("LEADING"));
+                ASSERT_TRUE(tester.getTester(1).waitForState("FOLLOWING"));
+                ASSERT_TRUE(tester.getTester(2).waitForState("FOLLOWING"));
```
Comment (rafecolton) [inline test/clustertest/tests/BadCommandTest.cpp:19]: What do these additions check for?

File: BedrockServer.cpp (hunk 1 above)
Comment (tylerkaraszewski) [reply BedrockServer.cpp:974]: Yes. This doesn't actually matter. The preexisting comment implied we thought this is what we had been doing the whole time.

Nothing should be getting `repeek`ed in the blocking queue regardless. We should only ever `repeek` anything before running `process`, and we only every move anything to the blocking queue after a commit conflict, which is inherently after `process` has happened.

Even if we did move a command back from the blocking to the main queue, it would just give it a few more opportunities to conflict before making it back to the blocking queue.

Also, the blocking queue is going to go away anyway as soon as this change is done.

File: BedrockServer.cpp (hunk 2 above)
Comment (tylerkaraszewski) [reply BedrockServer.cpp:2269]: It's not, this is just an artifact leftover from rearranging this block several times with different variations on it. I can change it back.

File: BedrockServer.cpp (hunk 4 above)
Comment (tylerkaraszewski) [reply BedrockServer.cpp:2288]: Here. Inserting it in the command queue handles it. [SScheduledPriorityQueue](https://github.com/Expensify/Bedrock/blob/main/libstuff/SScheduledPriorityQueue.h) doesn't dequeue commands that are scheduled in the future. They are enqueued behind any other commands that are scheduled earlier than them, and even if there are no commands in front of them, they're not dequeued until they're scheduled to run.

So, this maintains the behavior of pushing commands that are scheduled in the future (a feature we actually intend to remove eventually) into the command queue to be dequeued (and thus run) later.

This check is here to prevent them from being run immediately in `runCommand`, which is before they're scheduled.

File: test/clustertest/tests/BadCommandTest.cpp (hunk 5 above)
Comment (tylerkaraszewski) [reply test/clustertest/tests/BadCommandTest.cpp:19]: That the cluster is in the correct state. Mainly, that both followers are following, and we aren't in a state where leader is leading with one follower and a third server still starting up.

BadCommandTest is somewhat flaky and I think it's because we can crash a server before the last server is actually up and online, and then it doesn't handle everything quite as expected, so hopefully this makes this test a bit more reliable.

**Review by rafecolton — APPROVED** [review]
Discussed with @tylerkaraszewski 1:1, all makes sense now 👍🏼

File: BedrockServer.cpp [hunk 6]
```diff
@@ -2256,45 +2259,47 @@ void BedrockServer::handleSocket(Socket&& socket, bool fromControlPort, bool fro
...
+                                // These are implicitly fire-and-forget commands and a response will already have been sent in `buildCommandFromRequest`.
+                                _commandQueue.push(move(command));
+                                continue;
+                            } else {
+                                SINFO("Running new '" << command->request.methodLine << "' command from local client, with " << _commandQueue.size() << " commands already queued.");
```
Comment (roryabraham) [inline BedrockServer.cpp:2293]: Fun fact: renaming this log from `Queuing new` to `Running new` probably broke the "Auth Requests" graph in grafana, as we had a regex looking for that here: https://github.com/Expensify/Salt/blob/373a4551d7ef81489f91acc7346580d20a023047/shared_pkgs/bedrock/files/logsToStatsd.pl#L172-L189
